### PR TITLE
Fix allocatable assignment issues

### PR DESCRIPTION
**Pull request by jeanPerier**
- New storage for derived type requiring initialization was not initialized before doing the actual assignment, leading to crashes when the derived type had allocatable component which should have been initialized to a clean de-allocated status.
- Previous lower bounds of `alloc_array` where kept in things like `alloc_array = array + b`. This is wrong, they should be only kept if the right hand side is scalar (they are not even guaranteed to be well defined otherwise since `alloc_array` may be unallocated when the RHS is not a scalar).

### CODE REVIEW

File: flang/lib/Optimizer/Builder/MutableBox.cpp [hunk 1]
```diff
@@ -14,11 +14,66 @@
...
+/// Create a fir.box describing the new address, bounds, and length parameters
+/// for a MutableBox \p box.
+static mlir::Value createNewFirBox(fir::FirOpBuilder &builder,
+                                   mlir::Location loc,
+                                   const fir::MutableBoxValue &box,
+                                   mlir::Value addr, mlir::ValueRange lbounds,
+                                   mlir::ValueRange extents,
+                                   mlir::ValueRange lengths) {
+  mlir::Value shape;
+  if (!extents.empty()) {
```
Comment (psteinfeld) [inline flang/lib/Optimizer/Builder/MutableBox.cpp:32]: What happens if `extents.empty()` is true?

File: flang/lib/Optimizer/Builder/MutableBox.cpp [hunk 2]
```diff
@@ -14,11 +14,66 @@
...
+      }
+      auto shapeShiftType =
+          fir::ShapeShiftType::get(builder.getContext(), extents.size());
+      shape = builder.create<fir::ShapeShiftOp>(loc, shapeShiftType,
+                                                shapeShiftBounds);
+    }
+  }
+  if (addr.getType().isa<fir::BoxType>())
+    // The entity is already boxed.
+    return builder.createConvert(loc, box.getBoxTy(), addr);
```
Comment (psteinfeld) [inline flang/lib/Optimizer/Builder/MutableBox.cpp:51]: It looks like this code could be moved to the top of this funtion

**Review by psteinfeld — APPROVED** [review]
Aside from a couple of small comments, all builds and tests correctly and looks good.

File: flang/lib/Optimizer/Builder/MutableBox.cpp (hunk 1 above)
Comment (jeanPerier) [reply flang/lib/Optimizer/Builder/MutableBox.cpp:36]: The shape remains an empty mlir value, which is what emboxOp expects for scalars. I added a small comment to clarify.

**Pull request merged**